### PR TITLE
bpo-47012: speed up iteration of bytes and bytearray

### PR DESCRIPTION
**Pull request by kumaraditya303**
Benchmark:

```py
from pyperf import Runner, perf_counter

def bench_bytes(loops, length):
    src = b'helloworld' * length
    t0 = perf_counter()
    for _ in range(loops):
        for i in src:
            pass
    return perf_counter() - t0

def bench_bytearray(loops, length):
    src = bytearray(b'hello' * length)
    t0 = perf_counter()
    for _ in range(loops):
        for i in src:
            pass
    return perf_counter() - t0

runner = Runner()
for n in [10_000, 100_000]:
    runner.bench_time_func(f"bytes {n}", bench_bytes, n)
    runner.bench_time_func(f"bytearray {n}", bench_bytearray, n)
```

Results:
```console
bytes 10000: Mean +- std dev: [base] 829 us +- 38 us -> [patch] 677 us +- 44 us: 1.23x faster
bytearray 10000: Mean +- std dev: [base] 523 us +- 34 us -> [patch] 360 us +- 19 us: 1.45x faster
bytes 100000: Mean +- std dev: [base] 8.33 ms +- 0.38 ms -> [patch] 6.89 ms +- 0.75 ms: 1.21x faster
bytearray 100000: Mean +- std dev: [base] 5.19 ms +- 0.23 ms -> [patch] 3.61 ms +- 0.23 ms: 1.44x faster

Geometric mean: 1.33x faster
```

https://bugs.python.org/issue47012

### CODE REVIEW

File: Include/internal/pycore_long.h [hunk 1]
```diff
@@ -37,6 +37,11 @@ static inline PyObject* _PyLong_GetZero(void)
 static inline PyObject* _PyLong_GetOne(void)
 { return (PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS+1]; }
 
+static inline PyObject* _PyLong_FromUnsignedChar(unsigned char i)
+{
+    return Py_NewRef((PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS+i]);
```
Comment (thatbirdguythatuknownot) [inline Include/internal/pycore_long.h:43]: You might want to use `_Py_NewRef`, which is a `static inline` function, to make it faster.

Comment (kumaraditya303) [reply Include/internal/pycore_long.h:43]: There is no advantage as `Py_NewRef` is an alias of `_Py_NewRef`.

Comment (thatbirdguythatuknownot) [reply Include/internal/pycore_long.h:43]: > There is no advantage as `Py_NewRef` is an alias of `_Py_NewRef`.

I see.

Comment (ghost) [reply Include/internal/pycore_long.h:43]: Need to consider the case when no small integer cache is defined.
```c
#if defined(_PY_NSMALLPOSINTS) && _PY_NSMALLPOSINTS >= 256
    use_fast_code
#else
    use_fallback_code
#endif
```

Comment (thatbirdguythatuknownot) [reply Include/internal/pycore_long.h:42]: > Need to consider the case when no small integer cache is defined.

We don't need to. `pycore_runtime_init.h` already initializes all 5+257 longs whether or not `_PY_NSMALLPOSINTS` is >= 256.

**Comment by sweeneyde** [comment]
I agree with @animalize that it would be safest to include a preprocessor directive, whether that's with separate `#if _PY_NSMALLPOSINTS > 255`/`#else` code, or with a `#error` directive.

I'm not sure why anyone would compile with fewer than 256 cached small ints. [Here](https://github.com/python/cpython/commit/8e3b9f92835654943bb59d9658bb52e1b0f40a22), @vstinner added the code

```c
// _PyLong_GetZero() and _PyLong_GetOne() must always be available
#if _PY_NSMALLPOSINTS < 2
#  error "_PY_NSMALLPOSINTS must be greater than 1"
#endif
```

@vstinner, would there be any downside to requiring all of `(0, 1, ..., 255)` be in the small int cache?

File: Include/internal/pycore_long.h [hunk 2]
```diff
@@ -37,6 +37,11 @@ static inline PyObject* _PyLong_GetZero(void)
 static inline PyObject* _PyLong_GetOne(void)
 { return (PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS+1]; }
 
+static inline PyObject* _PyLong_FromUnsignedChar(unsigned char i)
+{
```
Comment (vstinner) [inline Include/internal/pycore_long.h:42]: You should update `#if _PY_NSMALLPOSINTS < 2` test above.

Comment (kumaraditya303) [reply Include/internal/pycore_long.h:42]: Fixed

**Review by sweeneyde — APPROVED** [review]

**Comment by kumaraditya303** [comment]
@sweeneyde I have added the compilation guard, but FYI if _PY_NSMALLPOSINTS is changed then it would break deepfreeze and module freezing infra so it is not configurable not to mention that it is declared in internal header.

**Pull request merged**